doc/adrs/065_add_tree_widget_in_forms.adoc

```diff
@@ -0,0 +1,111 @@
+A selectable node (e.g. one which represents a semantic element or representation) will change the workbench's selection when clicked on (using its `id`, `label` and `kind` attributes as `SelectionEntry`).
+Clicking on a non-selectable node will have no effect on the selection; in the UI the only effect will be to highlight it and toggle its expanded/collapsed state.
+
+The existing `Tree` *representation* is designed to be used directly through a subscription and manages its "expanded" state accordingly, with each change in the expansion depth creating a new representation and a new subscription.
+The new `TreeWidget` will be embedded in an enclosing `Form` representation.
+We do not want to invalidate and recompute the whole form when the user expands or collapses tree nodes.
+
+The `TreeWidget` will thus send its full content to the frontend, with the ids of the elements to expand for the initial display (the `TreeWidget.expandedNodesIds` attribute).
+Subsequent changes in the collapse/expand state of the nodes will be done purely on the client.
+If the underlying model changes, the frontend will receive a new state for the whole enclosing form, including the tree widget(s) anyway.
+
+Because the depth of a tree is not known beforehand by the frontend, we will use a flattened representation of the hierarchy so that the GraphQL subscription (on `FormRefreshedEventPayload`) has a fixed depth.
+Instead of recursive structure of `TreeItem` as used for the `Tree` representation, the `TreeWidget` contains a flat list of `TreeNode`, each with its `parentId`.
+Root nodes in the tree will have a null `parentId`.
+The frontend will have to rebuild the tree data structure from this flat list.
+
+=== Backend
+
+The backend representation of a tree will be plain POJOs corresponding to the `TreeWidget` and `TreeNode` GraphQL types above.
+As is the case for `TreeItem`, each concrete application (e.g. Sirius Web) will need a custom data fetcher to rewrite the `TreeNode.imageURL` attribute to make it relative to the server's image base path (see `TreeItemImageURLDataFetcher`).
+
+The backend API to describe a tree widget will follow the same pattern as the other `AbstractWidgetDescription`.
+It will have the following fields:
+
+- `Function<VariableManager, String> idProvider, labelProvider, iconURLProvider`: used to provide the unique id, label and optional icon URL of the actual `TreeWidget` being rendered.
+- `Function<VariableManager, List<?>> childrenProvider`: used to provide the ordered children of a particular candidate element.
+For a given render, this will be invoked *once* with both `self` and `candidate` variables set to the target element of the widget itself, and `ancestors` set to an empty list to determine the root elements to display.
+It will then be invoked recursively on every element returned with:
+  * `self` always set the target element of the widget itself;
+  * `candidate` set to each element in turn;
```

Comment:
It's very odd to use the name `candidate` here since the current element for the various providers / expressions has always been `self`. The name `candidate` has been introduced for the candidates in the select / radio widgets since it represents a candidate for the value.

Something like `root` to identify the root target element.